src/twisted/internet/tcp.py

```diff
@@ -853,6 +858,246 @@ def getPeer(self):
+
+    @_machine.state(initial=True)
+    def _unreserved(self):
+        """
+        The reservation file is unreserved.
+        """
+
+
+    @_machine.state()
+    def _reserved(self):
+        """
+        The reservation file is reserved.
+        """
+
+
+    @_machine.input()
+    def available(self):
+        """
+        Is the reservation available?
+        """
+
+
+    @_machine.input()
+    def release(self):
+        """
+        Release the reservation descriptor so another file can take
+        its place.
+        """
+
+
+    @_machine.input()
+    def acquire(self):
+        """
+        Ensure that the reservation file is held.
+        """
+
+
+    @_machine.input()
+    def replace(self, fileObject):
+        """
+        Replace a file with the reservation file only if our reserve
+        file is unclaimed.
+
+        @param fileObject: The file object to close.
+        """
+
+
+    @_machine.input()
+    def _openFailedWithEMFILE(self):
+        """
+        An internal input that undoes a reservation that failed to
+        open the reservation file.
+        """
+
+
+    def _createFD(self):
+        """
+        Actually open the reservation file.
+        """
+        try:
+            self._fileDescriptor = self._fileFactory()
+        except (IOError, OSError) as e:
+            if e.errno == EMFILE:
+                self._log.failure("Failed to open reservation file.")
+                self._fileDescriptor = None
+            else:
+                raise
+
+
+    @_machine.output()
+    def _openReservationFile(self):
+        """
+        Use the C{_fileFactory} to claim a file descriptor.
+        """
+        self._createFD()
+
+
+    @_machine.output()
+    def _reopenReservationFile(self, fileObject):
+        """
+        The same as L{_openReservationFile}, except that it accepts a
+        C{fileObject} argument.
+
+        @param fileObject: ignored.
+        """
+        self._createFD()
+
+
+    @_machine.output()
+    def _closeReservationFile(self):
+        """
+        Close the reservation file.
+        """
+        self._fileDescriptor.close()
+        self._fileDescriptor = None
+
+
+    @_machine.output()
+    def _closeFile(self, fileObject):
+        """
+        Close the provided file object.
+
+        @param fileObject: The file object to close.
+        """
+        fileObject.close()
+
+
+    @_machine.output()
+    def _returnTrue(self):
+        """
+        Return True.
+
+        @return: L{True}
+        """
+        return True
+
+
+    @_machine.output()
+    def _returnFalse(self):
+        """
+        Return False.
+
+        @return: L{False}
+        """
+        return False
+
+
+    @_machine.output()
+    def _replaced(self, fileObject):
+        """
+        Return True.
+
+        @param fileObject: ignored.
+
+        @return: L{True}
+        """
+        if self._fileDescriptor is None:
+            self._openFailedWithEMFILE()
+            return False
```

Comment:
This is wrong for two reasons: the file given to `replace` _was_ closed, so [user code can't use it](https://github.com/twisted/twisted/pull/996/files/d29f39116cf69586ab814a366acca24f64fddc1a#diff-d4edf4dd546357655acc84fff0ab7082R1324), and why on earth would something named `_returnTrue` ever return `False`.

Reply:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/124437a0814e35c6dafddfa21ec2f337960af5e1